Support fstype other then ext4

prateekpandey14 opened the pull request:
**What this PR does / why we need it:**
This commit will extend the provisioner to support other filesystem i.e. `XFS, ext3` etc. `Ext4` will be the default fstype for openebs volumes. But can be changed via storageclass parameter key/value for example `openebs.io/fstype : xfs`

Note: `OPENEBS_VALID_FSTYPE` var can be passed from openebs provisioner pod to support any new valid fstype and use it via StorageClass. for example `export OPENEBS_VALID_FSTYPE="xfs,ext3,btrfs" `

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: openebs-percona
  provisioner: openebs.io/provisioner-iscsi
parameters:
  openebs.io/storage-pool: "default"
  openebs.io/jiva-replica-count: "1"
  openebs.io/volume-monitor: "true"
  openebs.io/capacity: 5G
  openebs.io/fstype: "xfs"
   ```
fixes: #https://github.com/openebs/openebs/issues/1446
#https://github.com/openebs/openebs/issues/1454 

**How to verify this change ?**

Tested the changes : 
1. Local kubernets cluster with perconaDB 
2. GKE cluster with MongoDB

```
$ kubectl get po
NAME                                                           READY     STATUS    RESTARTS   AGE
maya-apiserver-76fdf7c6c5-mjcvj                                1/1       Running   0          55m
openebs-provisioner-77c69f5df6-54trd                           1/1       Running   0          55m
percona                                                        1/1       Running   0          4m
pvc-3244a601-4e2e-11e8-9997-54e1ad0c1ccc-ctrl-fd8898dc-l4h5l   1/1       Running   0          4m
pvc-3244a601-4e2e-11e8-9997-54e1ad0c1ccc-rep-df5d44bc-4k8rz    1/1       Running   0          4m
snapshot-controller-5bc67594d7-x49jc                           2/2       Running   0          55m

```
```sh
sudo df -Th /dev/sdb
Filesystem     Type  Size  Used Avail Use% Mounted on
/dev/sdb       xfs   5.0G  252M  4.8G   5% /var/lib/kubelet/pods/32442f55-4e2e-11e8-9997-54e1ad0c1ccc/volumes/kubernetes.io~iscsi/pvc-3244a601-4e2e-11e8-9997-54e1ad0c1ccc
```
```yaml
$ kubectl get pv pvc-3244a601-4e2e-11e8-9997-54e1ad0c1ccc -o yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    alpha.dashboard.kubernetes.io/links: '{"monitor":"http://127.0.0.1:32515/dashboard/db/openebs-volume-stats?orgId=1&var-OpenEBS=pvc-3244a601-4e2e-11e8-9997-54e1ad0c1ccc","maya":"https://mayaonline.io/"}'
    openEBSProvisionerIdentity: 127.0.0.1
    pv.kubernetes.io/provisioned-by: openebs.io/provisioner-iscsi
  creationTimestamp: 2018-05-02T17:28:16Z
  finalizers:
  - kubernetes.io/pv-protection
  name: pvc-3244a601-4e2e-11e8-9997-54e1ad0c1ccc
  resourceVersion: "7359"
  selfLink: /api/v1/persistentvolumes/pvc-3244a601-4e2e-11e8-9997-54e1ad0c1ccc
  uid: 32a7924a-4e2e-11e8-9997-54e1ad0c1ccc
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 5G
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: demo-vol1-claim
    namespace: default
    resourceVersion: "7328"
    uid: 3244a601-4e2e-11e8-9997-54e1ad0c1ccc
  iscsi:
    fsType: xfs
    iqn: iqn.2016-09.com.openebs.jiva:pvc-3244a601-4e2e-11e8-9997-54e1ad0c1ccc
    iscsiInterface: default
    lun: 0
    targetPortal: 10.0.0.42:3260
  persistentVolumeReclaimPolicy: Delete
  storageClassName: openebs-percona
status:
  phase: Bound

```
Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>

prateekpandey14 committed:
Support fstype other then ext4

This commit will extend the provisioner to support
other filesystem i.e. XFS, ext3 etc. Ext4 will be the default fstype
for openebs volumes. But can be changed via storageclass parameter
key/value for example "openebs.io/fstype : xfs"

---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: openebs-percona
  provisioner: openebs.io/provisioner-iscsi
parameters:
  openebs.io/storage-pool: "default"
  openebs.io/jiva-replica-count: "1"
  openebs.io/volume-monitor: "true"
  openebs.io/capacity: 5G
  openebs.io/fstype: "xfs"
    ---

Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>

prateekpandey14 committed:
Fix minikube start in travis

Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>

princerachit commented on openebs/openebs-provisioner.go:
```diff
@@ -164,6 +166,16 @@ func (p *openEBSProvisioner) Provision(options controller.VolumeOptions) (*v1.Pe
 		volAnnotations["alpha.dashboard.kubernetes.io/links"] = "{" + strings.Join(userLinks, ",") + "}"
 	}
 
+	fsType := ""
```
How about instead of traversing map we do:
```
if fsType=options.Parameter["openebs.io/fstype"]; fsType == ""{
    fsType = defaultFSType
}
```

princerachit replied:
We can validate fsType if we use a const map ^^^

prateekpandey14 replied:
Yes , if we have only 1 key to be be parsed , traversing of map is not required.

prateekpandey14 replied:
We will have some more key/value pairs related to volume policy as a different cases which can added here easily.

princerachit commented on openebs/openebs-provisioner.go:
```diff
@@ -43,6 +43,8 @@ const (
 	// It's currently still used and will be held for backwards compatibility
 	BetaStorageClassAnnotation = "volume.beta.kubernetes.io/storage-class"
+	//defaultFSType
+	defaultFSType = "ext4"
```
can we have a map of supported fsTypes so that we can validate if the requested type is valid or not?

prateekpandey14 replied:
Yes, we can have the map of supported fstype  once we validate them.

prateekpandey14 replied:
done.

kmova commented:
@prateekpandey14 -- doesn't the PVC already have a way of mentioning the format options? Is it defaulted to ext4?

prateekpandey14 commented:
No @kmova  PVC doesn't have a way of mentioning the format options, as i seen in PVC struct.

prateekpandey14 committed:
Add unit test for parsing storageclass parameters.

Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>

kmova commented on openebs/openebs-provisioner.go:
```diff
@@ -43,8 +43,13 @@ const (
 )
 
+// validFSType represents the valid fstype supported by openebs volume
+var validFSType = map[string]bool{"ext4": true, "xfs": true}
```
Can this valid list of options be passed as parameters to the pod? Extending this further, it will be good if all default options can be passed as configuration parameters to container.

prateekpandey14 replied:
Added `OPENEBS_VALID_FSTYPE` env variable which can be exported in provisioner POD 
to add fstype in a actual valid fstype supported list.

kmova commented on openebs/openebs-provisioner_test.go:
```diff
@@ -0,0 +1,43 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseClassParameters(t *testing.T) {
```
These tests can be simplified (remove the duplicate code) by following the TDD approach.

prateekpandey14 replied:
Done, changed to TDD approach.

prateekpandey14 committed:
Add OPENEBS_VALID_FSTYPE env as a config parameter

OPENEBS_VALID_FSTYPE can be exported as supported fstype for
openebs volume.
This can be set as ENV in openebs-provisioner pod deployment.
for example: OPENEBS_VALID_FSTYPE="ext3,btrfs,zfs"

Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>

princerachit commented on openebs/openebs-provisioner.go:
```diff
@@ -281,3 +291,41 @@ func GetStorageClassName(options controller.VolumeOptions) *string {
+	}
+
+	//Get openebs supported fstype from ENV variable
+	validENVFSType := os.Getenv("OPENEBS_VALID_FSTYPE")
```
Wondering if this could go into somewhere else like init block. Otherwise, this logic will run whenever there is a volume provision. The outcome will remain same in every execution.

utkarshmani1997 replied:
@princerachit i think this approach is good because user may require different FSTYPE for different  apps otherwise he needs to run another instance of openebs-provisioner in order to provision volume in the other requested format.Using this approach he can modify the env as per the requirement. 
@AmitKumarDas  @kmova what's your view on this ?

princerachit replied:
The `OPENEBS_VALID_FSTYPE` contains a comma separated supported fstypes. How is it going to change? The env var is intrinsic to openebs provisioner container. It has nothing to do with applications.

prateekpandey14 replied:
To modify `OPENEBS_VALID_FSTYPE` it in a running provisioner pod we have to use it `Getenv` here. If Admin wants to add any new fstype this will help here.

Putting in init() method only initialise once when pod starts, we can't change it while running.

princerachit replied:
Why/how would we modify `OPENEBS_VALID_FSTYPE` ?

prateekpandey14 replied:
How -- We can export it inside the provisioner pod` export 
OPENEBS_VALID_FSTYPE="zfs,abcd"`
Why -- see above.

princerachit replied:
Ok. So I had a discussion with @prateekpandey14 regarding the above issue. We are doing this to give the user(or the devops admin) control over what file systems could be created using openebs. This has an obvious caveat that wrong or invalid values in `OPENEBS_VALID_FSTYPE` will break formatting. However, the onus to ensure that correct values are passed as well as used lies on the devops admin. Approving this PR now.

prateekpandey14 commented:
Makefile doesn't run the test in travis  so here is the test results.
```sh
go test -v
=== RUN   TestParseClassParameters
=== RUN   TestParseClassParameters/Dafault_fstype
=== RUN   TestParseClassParameters/ext4_fstype
=== RUN   TestParseClassParameters/xfs_fstype
=== RUN   TestParseClassParameters/Invalid_fstype
--- PASS: TestParseClassParameters (0.00s)
    --- PASS: TestParseClassParameters/Dafault_fstype (0.00s)
    --- PASS: TestParseClassParameters/ext4_fstype (0.00s)
    --- PASS: TestParseClassParameters/xfs_fstype (0.00s)
    --- PASS: TestParseClassParameters/Invalid_fstype (0.00s)
=== RUN   TestParseClassENVParameters
=== RUN   TestParseClassENVParameters/Dafault_fstype
=== RUN   TestParseClassENVParameters/ext4_fstype
=== RUN   TestParseClassENVParameters/ENV_fstype_nfs
=== RUN   TestParseClassENVParameters/Invalid_fstype
=== RUN   TestParseClassENVParameters/ENV_fstype_btrfs
=== RUN   TestParseClassENVParameters/ENV_fstype_zfs
--- PASS: TestParseClassENVParameters (0.00s)
    --- PASS: TestParseClassENVParameters/Dafault_fstype (0.00s)
    --- PASS: TestParseClassENVParameters/ext4_fstype (0.00s)
    --- PASS: TestParseClassENVParameters/ENV_fstype_nfs (0.00s)
    --- PASS: TestParseClassENVParameters/Invalid_fstype (0.00s)
    --- PASS: TestParseClassENVParameters/ENV_fstype_btrfs (0.00s)
    --- PASS: TestParseClassENVParameters/ENV_fstype_zfs (0.00s)
PASS
ok      github.com/kubernetes-incubator/external-storage/openebs    0.007s
```

kmova reviewed (approved):
LGTM
